src/coreclr/vm/assembly.cpp

```diff
@@ -617,10 +624,22 @@ Assembly *Assembly::CreateDynamic(AppDomain *pDomain, CreateDynamicAssemblyArgs
             // Once everything is setup and nothing can fail anymore, the ownership will be
             // atomically transfered by call to LoaderAllocator::ActivateManagedTracking().
             pAssemblyLoaderAllocator->SetupManagedTracking(&args->loaderAllocator);
+            createdNewAssemblyLoaderAllocator = TRUE;
+
+            pLoaderAllocator = pAssemblyLoaderAllocator;
         }
         else
         {
-            pLoaderAllocator = pDomain->GetLoaderAllocator();
+            if (pBinderContext != nullptr)
+            {
+                pBinderContext->GetLoaderAllocator((LPVOID*)&pAssemblyLoaderAllocator);
+            }
+
+            pLoaderAllocator = pAssemblyLoaderAllocator == nullptr ? pDomain->GetLoaderAllocator() : pAssemblyLoaderAllocator;
+        }
+
+        if (~createdNewAssemblyLoaderAllocator)
```

Comment:
Nit: I don't think we use the `~` as a "not boolean" operator anywhere in the code base. It's probably technically the more correct one, but I think we rely on simple `!` everywhere for now.

Reply:
Replaced with `!`